analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/CompoundAssignment.cs

```diff
@@ -51,6 +52,29 @@ private static ProgramState ProcessNumericalCompoundAssignment(ProgramState stat
         }
     }
 
+    private static ProgramState ProcessDelegateCompoundAssignment(ProgramState state, ICompoundAssignmentOperationWrapper assignment)
+    {
+        // When the -= operator is used on a delegate instance (for unsubscribing),
+        // it can leave the invocation list associated with the delegate empty. In that case the delegate instance will become null.
+        // https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/subtraction-operator#delegate-removal
+        // For this reason, we remove the NotNull constraint from the delegate instance.
+        if (assignment.Target.Type.TypeKind == TypeKind.Delegate
+            && assignment.OperatorKind == BinaryOperatorKind.Subtract
+            && state.HasConstraint(assignment.Target, ObjectConstraint.NotNull))
+        {
+            var value = (state[assignment.Target] ?? SymbolicValue.Empty).WithoutConstraint(ObjectConstraint.NotNull);
+            state = state
+                .SetOperationValue(assignment, value)
+                .SetOperationValue(assignment.Target, value);
+            state = state.SetSymbolValue(assignment.Target.TrackedSymbol(state), value);
```

Comment:
Note: the left side of the -= operator must be a field, property, or indexer. In the first two cases, the symbol is always available. For indexers, I haven't been able to create a test case where the indexer would have `NotNull` as a constraint.
e.g.
```cs
this[42] = delegate { };
this[42] += Callback; // this[42] doesn't have the NotNull constraint
this[42] -= Callback; 
```
So I don't check whether the tracked symbol is `null`.